Article: What is the ESP-Registry (#175)
espressif/developer-portal: content/blog/2024/10/what-is-the-esp-registry/index.md

---
title: "What is the ESP-Registry?"
date: 2024-10-27
showAuthor: false
authors:
  - "pedro-minatel"
tags: ["Registry", "Component", "ESP-IDF"]
---

When starting a new project, we often reuse code from existing components, such as libraries and drivers, to save time. This not only saves time but also makes your project easier to maintain.

> In this article, we will explore what the ESP-Registry is and the key concepts of components, making it easy to understand not only how to use components but also how to contribute.

Reviewer: @kumekay 

There is a [Jira ticket](https://jira.espressif.com:8443/browse/PACMAN-842) to make the name consistent. Its status is done, but I still don't see the conclusion about which name is correct.

> In this article, we will explore what the ESP-Registry is and the key concepts of components, making it easy to understand not only how to use components but also how to contribute.

Reviewer 2: > @kumekay
> 
> There is a [Jira ticket](https://jira.espressif.com:8443/browse/PACMAN-842) to make the name consistent. Its status is done, but I still don't see the conclusion about which name is correct.

For some reason, MRs weren't automatically linked to the Jira ticket. I added them to a comment.

The chosen variant is ESP Component Registry,  

We may consider dropping the word  “Component”, but then the domain name "components.espressif.com" need to be updated, and it's a non-trivial technical task.

## Concept

> In the [ESP-IDF](https://idf.espressif.com), components are widely used to make the framework more modular, making it clearer to understand and maintain. The naming convention in the ESP-IDF is to use the term "component" instead of "library" because the ESP-IDF is more like an operating system with self-contained modules that interact with the system at a higher level.

Reviewer: ```suggestion
The architecture of [ESP-IDF](https://idf.espressif.com) is modular, making it easier to understand and maintain. Such modules are called *components*, as opposed to *libraries*, because they are self-contained and used on top of the ESP-IDF core.
```
Suggest not using the term operating system, because it implies that the system is run on top of the hardware and interacts with it. ESP-IDF is not run on hardware, it is a framework that is used to generate applications which are run on hardware.

Author: Using the word "operating system like" is common for the ESP-IDF. But we can avoid it here and open another discussion about it.

> There are three concepts that we need to introduce in order to understand the complete solution:

Reviewer: ```suggestion
There are three concepts that we need to introduce in order to understand the complete solution: component, component manager, and component register. Let's go through them one by one.
```

### 1. Component

> As mentioned before, the component is a self-contained piece of code that can be used to add a new functionality to the application. In most of the cases, the component is used in several applications, making easy to reuse code and as well will improve the efficiency and reduce the development complexity.

Reviewer: ```suggestion
As mentioned above, a component is a self-contained piece of code that contains new functionality and can be added to an application. In most cases, components are used in multiple applications, making it easy to reuse code, improve the efficiency, and reduce development complexity.
```

> For example, in your application (`my-application`), you are going to use a specific sensor (`sensor-component`) that will collect temperature and humidity from the environment. Instead of adding all the logic behind the communication protocol, data collection, conversions and etc to the application code, you can add as a component.

Reviewer: ```suggestion
For example, let's imagine there is an application called `my-application` in which we want to use a specific sensor that will collect temperature and humidity from the environment. Instead of adding all the logic behind the communication protocol, data collection, conversions, etc. to the application code, you can add a component `sensor-component`, that has this logic already implemented in code. The important part is that this code has been reviewed and tested multiple times by other developers. 
```

```text
.
└── my-application
    ├── components
    │   └── sensor-component
    │       └── include
    │           └── sensor-component.c
    │       └── src
    │           └── CMakeLists.txt
    │           └── sensor-component.c
    │       └── CMakeLists.txt

>     ├── CMakeLists.txt

Reviewer 2: This application layout is not a typical one, usually we have a `main` component where the code is contained:

```
    ├── CMakeLists.txt
    ├── main
    │   └── CMakeLists.txt
    │   └── my-application.c
    |   └── include
    │       └── my-application.h
    └── my-application.c
```

Author: One more thing. Should we create the `src` folder for the component?
Currently, the `idf.py create-component` does not create the `scr` folder.

Reviewer 2: I think 'src' directory is optional, most simple components inside ESP-IDFs components directory do not have it and source files are directly in the component directory.

However, I want to update "create-component", "create-project" and 'create-manifest' templates because they are too minimalistic and a bit outdated now.

Author: I'll remove this part from this article and add it to the new article "Creating Components".

    ├── include
    │   └── my-application.h
    └── my-application.c
```

> In this example, the `sensor-component` will be added to the project as a component, and can be shared as is among different projects that uses the same sensor.

Reviewer: ```suggestion
In this example, the `sensor-component` has been added to the project as a component. This component can be shared as is among different projects that use the same sensor.
```

### 2. Component Manager

> To improve the developer experience, Espressif has developed a set of tools to help manage the component inside the project. The Component Manager is a CLI tool that allows you to create new components and add existing components to your project.

Reviewer: ```suggestion
Single components can quickly go out of hand if not managed properly. The component manager does just that -- it is a CLI tool that allows you to add existing components to your project and create new components.
```

#### Creating new components

> To create a new component using the component manager, you can use:

Reviewer: As we discussed it, maybe it is worth dropping this subsection, because this information is not sufficient on its own. It is better to have it in a how to guide with more details about how to create a component.

> To create a new component using the component manager, you can use:

Reviewer 2: As an alternative, the command can be:

`idf.py create-component -C ./components component-name`

> To create a new component using the component manager, you can use:

Author: Let's keep this to the next article.

```bash
idf.py create-component component-name
```

By using this command, all the necessary files will be included in the component folder. This command should be executed inside the `components` folder.

### 3. Component Registry

<figure style="width: 90%; margin: 0 auto; text-align: center;">
    <img
        src="./assets/cm_logo.webp"
        alt="ESP-Registry Logo"
        title="ESP-Registry Logo"
        style="width: 100%;"
    />
    <figcaption>ESP-Registry Logo</figcaption>
</figure>

> To support the high number of components in the ESP-IDF, [Espressif](https://espressif.com) has developed the [ESP-Registry](https://components.espressif.com/), a platform that keeps all components available for anyone to use in a central repository. With the ESP-Registry, users can easily find, install, use examples of, and eventually create and upload their own components.

Reviewer: ```suggestion
Now we have lots of components that can be managed in a project. But to make these components accessible to everyone, there should be some kind of central repository. This role is played by the ESP Component Registry. It helps maintain ESP-IDF components for users to easily find, install, use examples of, and eventually create and upload their own components.
```

<figure style="width: 90%; margin: 0 auto; text-align: center;">
    <img
        src="./assets/esp-registry-main.webp"
        alt="ESP-Registry Landing Page"
        title="ESP-Registry Landing Page"
        style="width: 100%;"
    />
    <figcaption>ESP-Registry Landing Page</figcaption>
</figure>

> Currently, the Registry has **+500 components**, developed by Espressif, partners, and the community.

Reviewer: ```suggestion
Currently, the Registry has **500+ components**, developed by Espressif, partners, and the community.

```

Some examples:

- [led_strip](https://components.espressif.com/components/espressif/led_strip/): Driver designed for the addressable LEDs like WS2812.
- [ds18b20](https://components.espressif.com/components/espressif/ds18b20/): Driver for the DS18B20 temperature sensor using the 1-Wire protocol.
- [arduino-esp32](https://components.espressif.com/components/espressif/arduino-esp32): Component for the Arduino core for the ESP32 to be used inside the ESP-IDF.

> To get started with the ESP-Registry, you will need just to search for a component and add it to your project. If you want to publish your component to the registry, you will need an GitHub account.

Reviewer: ```suggestion
To get started with the ESP-Registry, just search for a component and add it to your project. If you want to publish your component to the ESP-Registry, you will also need a GitHub account.
```

- **Modularity**: The component will reduce the application complexity.
- **Flexibility**: Each component can be configured using individual KConfig.
- **Test**: Each component can be tested individually.

Using your own components or third-party components can speed up your development and it will be easy to maintain the application that uses the components.

One example we can use is a bug in a specific component that was fixed in the latest version. You can run a command that will update the component to the latest version automatically.

## Why should I write my own component?

Right now, there is probably a developer searching for some driver, library, or anything that could help them develop their project. If you have what they are looking for, but you haven't shared it anywhere, you will not be able to know and help them. Don't forget that it could be you next time.

Besides that, making components can also help you with other projects that might use the same components, and keeping them in the Registry can save your time looking where the component was stored and how you manage the version, download, etc.

> The main objective for the ESP-Registry is to keep a single platform for components, where developers across the world can search and use the shared components, however, the way the components are managed by the Component Manager, makes the use of components very easy for any developer. The components are managed in a way that you can specify the components in a `.yml` file as a dependency for your project and the build system will download from the Registry the component automatically during the first build.

Reviewer 2: > as a dependency for your project

I think we need to be more precise and say "as a dependency for any component in your project" because having many idf_component.yml files is a very common case.

> > Sharing knowledge is one of the most rewarding things you can do.

Author: This quote is not necessary, maybe removing will be better.

You can find more components examples at the [ESP-BSP](https://github.com/espressif/esp-bsp/tree/master/components) and the [esp-iot-solution](https://github.com/espressif/esp-iot-solution) repository.

## Conclusion

> Using components in your project can not only save development time but also creates the project modularity and the components can be tested individually. The ESP-Registry can be a resource for parts you do not need to develop yourself.

Reviewer: ```suggestion
Using components in your project not only saves development time but also results in project modularity and the components can be tested individually. The ESP-Registry can be a resource for parts you do not need to develop yourself.
```

---
title: "What is the ESP Component Registry?"
date: 2024-10-27
showAuthor: false
authors:
  - "pedro-minatel"
tags: ["Registry", "Component", "ESP-IDF"]
---

When starting a new project, we often reuse existing code, for example from libraries and drivers. This not only saves time but also makes your project easier to maintain.

In this article, we will explore what the [ESP Component Registry](https://components.espressif.com) is and the key concepts of components, making it easy to understand not only how to use components but also how to contribute.

## Concept

The architecture of [ESP-IDF](https://idf.espressif.com) is modular, making it easier to understand and maintain. Such modules are called *components*, as opposed to *libraries*, because they are self-contained and can also communicates with other components. Components are used on top of the ESP-IDF core.

There are three concepts that we need to introduce in order to understand the complete solution: component, component manager, and component register. Let's go through them one by one.

### 1. Component

> As mentioned above, a component is a self-contained piece of code that contains new functionality and can be added to an application. In most cases, components are used in multiple applications, making it easy to reuse code, improve the efficiency, and reduce development complexity.

Reviewer: ```suggestion
As mentioned above, a component is a self-contained piece of code that offers new functionality and can be added to an application. In most cases, components are used in multiple applications, making it easy to reuse code, improve the efficiency, and reduce development complexity.

```

For example, let's imagine there is an application called `my-application` in which we want to use a specific sensor that will collect temperature and humidity from the environment. Instead of adding all the logic behind the communication protocol, data collection, conversions, etc. to the application code, you can add a component i.e `sensor-component`, that has this logic already implemented in code. The important part is that this code has been reviewed and tested multiple times by other developers.

> The component can be created directly in the application, inside a specific folder, named `components` or can be created in a separated component project and this component can be added as a dependency to your application. Both solutions works but the recommended way is to create a new component project and add it as a dependency.

Reviewer: ```suggestion
The component can be either created directly in the application, inside a specific folder named `components`, or it can be created in a separated component project and then added as a dependency to the application. Both solutions work but the recommended way is to create a new component project and add it as a dependency.
```

### 2. Component Manager

Single components can quickly go out of hand if not managed properly. The component manager does just that -- it is a CLI tool that allows you to add existing components to your project and create new components.

With the component manager you can:

- Add a component as a dependency to your application .
- Create new components structure.
- Downloads the dependencies automatically when running CMake.

> After adding a component as a dependency to your project and running the build process, the component manager will download the component to your project in a folder named `managed_components`. All the dependencies will be downloaded directly from the ESP Component Registry repository and it will be stored inside this folder.

Reviewer: Misplaced *it*. It this edit correct?

```suggestion
After adding a component as a dependency to your project and running the build process, the component manager will download the component to your project in a folder named `managed_components`. All the dependencies will be downloaded directly from the ESP Component Registry repository and will be stored inside this folder.
```

- [esp_insights](https://components.espressif.com/components/espressif/esp_insights/versions/1.2.2): Remote diagnostics solution that allows users to remotely monitor the health of ESP devices in the field.

To get started with the ESP Component Registry, just search for a component and add it to your project. If you want to publish your component to the Registry, you will also need a GitHub account.

<figure style="width: 100%; margin: 0 auto; text-align: center;">
    <img
        src="./assets/component-screen.webp"
        alt="Component"
        title="Component Page"
        style="width: 100%;"
    />
    <figcaption>Component Page</figcaption>
</figure>

The process for publishing a component will be described in another article.

#### Adding components from ESP Component Registry

For adding a component as a dependency to your project, you can use:

```bash
idf.py add-dependency "espressif/led_strip^3.0.0"
```

or you can create a new project from the example:

```bash
idf.py create-project-from-example "espressif/led_strip^3.0.0:led_strip_rmt_ws2812"
```

> After running the command, you will need to build the application to download the components to the `managed_folder`.

Reviewer 2: ```suggestion
After running the command, you will need to build the application to download the components to the `managed_components` folder.

```

If the component is not added to your project after running build command, you can try to run the `reconfigure` command and build again:

```bash
idf.py reconfigure
```

## Advantages of using components

Components are widely used in the ESP-IDF and many other applications based on the ESP-IDF.

The main advantages of using components are:

- **Code reuse**: You can reuse code among different applications by sharing the component.
- **Modularity**: The component reduce the application complexity.

- **Flexibility**: Each component can be configured using individual Kconfig.
- **Test**: Each component can be tested individually.

Using your own components or components developed by Espressif, other companies, or the community, can speed up your development and it will be easy to maintain the application that uses the components.

> One example we can use is a bug in a specific component that was fixed in the latest version. You can run a command that will update the component to the latest version automatically.

Reviewer: ```suggestion
One example we can use is dealing with a bug in a specific component that has been fixed in the latest version. You can run a command that will update the component to the latest version automatically.

```

## Advantages of writing your own components

> Right now, there is probably a developer searching for some driver, library, or anything that could help them develop their project. If you have what they are looking for, but you haven't shared it anywhere, you will not be able to know and help them. Don't forget that it could be you next time.

Reviewer: ```suggestion
At this very moment, there might be a developer somewhere in the world searching for a driver, library, or anything that could help them develop their project. If you have what they are looking for but you haven't shared it anywhere, they will have to reinvent the wheel and write it themselves. Next time you can end up in the same situation.
```

You can write many different types of components for your projects, including:

- Drivers
- Libraries
- Protocols
- Encoders/Decoders
- Parsers
- Features
- Utilities
- Tools
- Frameworks
- HAL
- Board Support Packages

Besides that, making components can also help you with other projects that might use the same components, and keeping them in the Registry can save your time looking where the component was stored and how you manage the version, download, etc.

The main objective for the ESP Component Registry is to keep a single platform for components, where developers across the world can search and use the shared components, however, the way the components are managed by the Component Manager, makes the use of components very easy for any developer. The components are managed in a way that you can specify them as dependencies for your project and the build system will download the components from the Registry automatically during the first build.

You can find more components examples to inspire you at the [ESP-BSP](https://github.com/espressif/esp-bsp/tree/master/components) and the [esp-iot-solution](https://github.com/espressif/esp-iot-solution) repository.

## Conclusion

> Using components in your project not only saves development time but also results in project modularity and the components can be tested individually. The ESP Component Registry can be a resource for parts you do not need to develop yourself.

Reviewer: ```suggestion
Using components in your project not only saves development time but also results in project modularity, which allows testing components individually. The ESP Component Registry can be a resource for parts you do not need to develop yourself.

```
